### ISSUE TITLE
Batch size picking and corresponding learning rate on custom dataset.

### ISSUE THREAD

**Issue by wxthss82** [zylo117/Yet-Another-EfficientDet-Pytorch#156]
@zylo117  Thank you for your great repos. It's one of the best open source repos I've ever seen. So I read all closed issue before I do experiments with my dataset. Besides above, I have three questions and one suggestion: 

### Three Questions:
1. I can train my custom datasets with:
export CUDA_VISIBLE_DEVICES=0,1,2,3;python3 train.py -c 0 --batch_size 128 --lr 1e-5 --num_epochs 50  --load_weights logs/xxxx/efficientdet-d0_17_15500.pth  --head_only True -p xxxx --num_workers 20 --debug True  --val_interval 5 --es_patience 10

and GPU usage:

|   0  GeForce GTX 108...  Off  | 00000000:04:00.0 Off |                  N/A |
| 28%   51C    P2   221W / 250W |   5011MiB / 11178MiB |     56%      Default |
+-------------------------------+----------------------+----------------------+
|   1  GeForce GTX 108...  Off  | 00000000:05:00.0 Off |                  N/A |
| 28%   52C    P2   215W / 250W |   5009MiB / 11178MiB |    100%      Default |
+-------------------------------+----------------------+----------------------+
|   2  GeForce GTX 108...  Off  | 00000000:06:00.0 Off |                  N/A |
| 26%   48C    P2   202W / 250W |   5009MiB / 11178MiB |     99%      Default |
+-------------------------------+----------------------+----------------------+
|   3  GeForce GTX 108...  Off  | 00000000:07:00.0 Off |                  N/A |
| 27%   50C    P2   208W / 250W |   5009MiB / 11178MiB |    100%      Default |

Is the batch size 128 ok? if it is, Is the corresponding learning rate 1e-5 ok? I 

2. I trained about half day, The loss log shows like below:

![image](https://user-images.githubusercontent.com/3448682/79798316-6831ac00-838b-11ea-88de-2b6415812a7c.png)

The classification loss doesn't decease much. subquestion:
(a). Should I wait for long time keeping watch loss's changes( if so, can you make a suggestion about how much time it takes for loss to reach a low level) 
(b). I should using a smaller lr or smaller batch-size?
(c). Should I follow instructions prompt in other issue(Train head first, see if the performance is ok. If not, unfreeze the backbone and resume training.)

I'm doing (a) and (c) now.

3. About anchor config generation.

I'm using default coco anchor config for my dataset and I see

> Thanks very much. I forgot to change back to the default anchor ratio while inferring. Now everything is good. Thanks for the excellent work! Great job!

So the default anchor maybe suitable for my dataset. But I want to calc my anchor config. I see [https://github.com/zylo117/Yet-Another-EfficientDet-Pytorch/issues/27#issuecomment-612550889](url)

> Find out 9 anchors ratio with kmeans.
2. Decompose it into two 3x1 arrays, one is anchors_scales, the other is anchors_ratios_raw.
3. Decompose anchors_ratios_raw into 3x2 (width/height factor) and 3x1 arrays (base factor). In this case, it's [[1,1], [2,1], [1,2]] and [1, 0.7, 0.7]

But I don't get the point of step 2 and 3, nor does my manager. Can you give more hit on this?

Thank you very much, the great repo makes me learn a lot.

### One advice:
You can specify some issue requirements such at env, training cmd, train log when people open an issue.

**Comment by kaishijeng** [comment]
I have a similar problem which cls_loss is stuck with 0.2 range and not decreasing after 20th epoch,
What value of learning rate should I use to avoid this?

**Comment by zylo117** [comment]
@wxthss82 
1. I think your lr is a little bit too low, so the it takes more time.
2. The training seems ok, you should keep waiting.
3. anchor decomposition is quite complicated and it requires prior knowledge or experience to tune it. You don't actually need to tune it to get a good result as long as your boxes shapes are similar to coco's.

**Comment by zylo117** [comment]
@kaishijeng try raising lr if it's not overfitting. or switch to sgd, using `--optim sgd`

**Comment by wxthss82** [comment]
> I think your lr is a little bit too low, so the it takes more time.
The training seems ok, you should keep waiting.
anchor decomposition is quite complicated and it requires prior knowledge or experience to tune it. You don't actually need to tune it to get a good result as long as your boxes shapes are similar to coco's.

@zylo117 Thank you very much, I'll follow your instructions. 
One things I forget to clarify is that, before I using 1e-5 lr to do the training, I've trained about 40 epochs using lr 1e-4. when the loss doesn't seem to decrease (around 0.2 as @kaishijeng metioned), I think it's time to change the lr to 1e-5. Do you think it's ok? Or I'm not patient that I should wait for more time to see if using lr 1e-4 that the loss can decrease? Or using lr 1e-5 keep training?

Previous using lr 1e-4 command as follows:
python3 train.py -c 0 --batch_size 256 --lr 1e-4 --num_epochs 50  --load_weights logs/xxxx/efficientdet-d0_1_197.pth  --
head_only True -p xxxx --num_workers 36 

tep: 16307. Epoch: 35/50. Iteration: 453/453. Cls loss: 0.19615. Reg loss: 0.02343. Total loss: 0.21959: 100%|█████████████████████████████████████████████████████████████| 453/453 [35:02<00:00,  4.64s/it]
Val. Epoch: 35/50. Classification loss: 0.18747. Regression loss: 0.02428. Total loss: 0.21176
Step: 16499. Epoch: 36/50. Iteration: 192/453. Cls loss: 0.19282. Reg loss: 0.02508. Total loss: 0.21790:  42%|█████████████████████████▋                                   | 191/453 [16:04<22:06,  5.06s/it]ch
eckpoint...
Step: 16760. Epoch: 36/50. Iteration: 453/453. Cls loss: 0.19684. Reg loss: 0.02588. Total loss: 0.22272: 100%|█████████████████████████████████████████████████████████████| 453/453 [34:25<00:00,  4.56s/it]
Val. Epoch: 36/50. Classification loss: 0.18588. Regression loss: 0.02424. Total loss: 0.21012
Step: 16999. Epoch: 37/50. Iteration: 239/453. Cls loss: 0.19360. Reg loss: 0.02576. Total loss: 0.21935:  53%|████████████████████████████████                             | 238/453 [17:58<07:30,  2.10s/it]ch
eckpoint...
Step: 17213. Epoch: 37/50. Iteration: 453/453. Cls loss: 0.20758. Reg loss: 0.02805. Total loss: 0.23563: 100%|█████████████████████████████████████████████████████████████| 453/453 [33:12<00:00,  4.40s/it]
Val. Epoch: 37/50. Classification loss: 0.18447. Regression loss: 0.02419. Total loss: 0.20866
Step: 17499. Epoch: 38/50. Iteration: 286/453. Cls loss: 0.17894. Reg loss: 0.02337. Total loss: 0.20231:  63%|██████████████████████████████████████▍                      | 285/453 [14:18<06:28,  2.31s/it]ch
eckpoint...
Step: 17666. Epoch: 38/50. Iteration: 453/453. Cls loss: 0.18776. Reg loss: 0.02570. Total loss: 0.21346: 100%|█████████████████████████████████████████████████████████████| 453/453 [20:34<00:00,  2.73s/it]
Val. Epoch: 38/50. Classification loss: 0.18342. Regression loss: 0.02421. Total loss: 0.20763
Step: 17960. Epoch: 39/50. Iteration: 294/453. Cls loss: 0.18200. Reg loss: 0.02429. Total loss: 0.20628:  65%|███████████████████████████████████████▌                     | 294/453 [23:03<16:01,  6.05s/it]


Step: 17962. Epoch: 39/50. Iteration: 296/453. Cls loss: 0.18127. Reg loss: 0.02197. Total loss: 0.20324:  65%|███████████████████████████████████████▊                     | 296/453 [23:07<10:20,  3.95s/it]

**Comment by zylo117** [comment]
So the loss drops when lr is reduced? Keep training. 
But then again, you should run coco_eval to keep track of its mAP once in a while. It might have converged already

**Comment by wxthss82** [comment]
At the end of training with lr 1e-4, the loss reached 0.18. when switch lr to 1e-5, the loss randomized at 0.23-0.29 (current time  training statistics). 
I use coco_val, mAP doesn't change much.

 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.204
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.426
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.162
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.121
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.294
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.159
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.364
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.396
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.217
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.542

**Comment by zylo117** [comment]
Can you share the loss graph, I have a feeling that this is the best the model can do. I guess your dataset is quite difficult to learn.

**Comment by wxthss82** [comment]
I don't think it's difficult, causing our previous algorithm can reach high mAP. EfficientDet would be better. Or should I try to use EfficientDet-D2? But I want to see if D0 can only achieve this.
![image](https://user-images.githubusercontent.com/3448682/79832063-33e5dc00-83db-11ea-8646-14010cc7efdd.png)

**Comment by zylo117** [comment]
Why isn't there a val loss?

**Comment by wxthss82** [comment]
Sorry, here:

![image](https://user-images.githubusercontent.com/3448682/79837102-8925eb80-83e3-11ea-9425-aa1e9673299d.png)
![image](https://user-images.githubusercontent.com/3448682/79837119-8fb46300-83e3-11ea-81da-161610443d83.png)

**Comment by zylo117** [comment]
I see. Did you try `--head_only False`

**Comment by wxthss82** [comment]
I tried but it didn't output tensorboard info, I'll retry.

**Comment by wxthss82** [comment]
The graph is based on --head_only False, the training process with --head_only True didn't output tensorboard info, I'll retrain and give loss graph with --head_only True
